### ISSUE TITLE
Incorrect aggregative counts for bio.tools

### ISSUE THREAD

**Issue by bryan-brancotte** [edamontology/edam-browser#45]
Thanks to @matuskalas  in https://github.com/edamontology/edam-browser/issues/20#issuecomment-769239707 who spot that the count of  https://edamontology.github.io/edam-browser/#operation_3198 is wrong for bio.tools. 

I think it might be related to the fact that the term is duplicated, and thus counted twice, but when looking at https://edamontology.github.io/edam-browser/#topic_0084 the descendant count make no sens at all for now.

**Comment by matuskalas** [comment]
Indeed, thanks for creating a dedicated bug issue for it @bryan-brancotte ! 👍🏽

Are the numbers returned by the bio.tools API calls correct in these cases?

**Comment by bryan-brancotte** [comment]
I think API responses are ok (quick test indicat it) it is the aggregation of the descendant made here.

If bio.tools API would return the count including the descendants, it would be both safer as unit tested, and faster ! ;)

**Comment by HagerDakroury** [comment]
Hi, can I get some guidance for testing this?

As I understood, the count should be the current term count + the descendants' counts.

The error offset for [#operation_3198](https://edamontology.github.io/edam-browser/#operation_3198) is almost double the current value. But for [#topic_0084](https://edamontology.github.io/edam-browser/#topic_0084) it's ~100. So the duplication being the issue is ruled out for now.

Is this correct?

**Comment by matuskalas** [comment]
Thank you a billion times @HagerDakroury for looking into this issue! 👍🏽🙏🏽

I found an example with smaller numbers: https://edamontology.github.io/edam-browser/#data_2977. For **_as output_**: 20 times as is, 80 times with 3 "descendants" (wrong), the 3 sub-concepts have 15, 13, and 6. So the sum should have been <=54

**Comment by matuskalas** [comment]
Another: https://edamontology.github.io/edam-browser/#format_1974
1 time _**as output**_, 66 times with 1 "descendant", in fact the only sub-concept is _not used_ 🤔

**Comment by HagerDakroury** [comment]
> Thank you a billion times @HagerDakroury for looking into this issue! 👍🏽🙏🏽
> 
> I found an example with smaller numbers: https://edamontology.github.io/edam-browser/#data_2977. For **_as output_**: 20 times as is, 80 times with 3 "descendants" (wrong), the 3 sub-concepts have 15, 13, and 6. So the sum should have been <=54

That definitely seems pretty random. Even if sub-concepts are counted twice (duplicate). The maths doesn't add up!

> Another: https://edamontology.github.io/edam-browser/#format_1974
> 1 time _**as output**_, 66 times with 1 "descendant", in fact the only sub-concept is _not used_ 🤔

hmm, I'm not getting 66 times? Here's what comes up
![image](https://user-images.githubusercontent.com/37930475/113755080-2fabef00-9710-11eb-8e89-6709abfa7e21.png)

Actually, that example should have _not used as input  (1) - 1 times as output (1)_ . So maybe the sub-concept usage as input got mistakingly counted as an output?

**Comment by matuskalas** [comment]
Indeed, now I'm getting the same as, with 2. Maybe I mixed it up yesterday.

In any case, I found out where the numbers come from! 🙌🏽
They are just the actual number for the given concept, multiplied by the number of its sub-concepts +1 (the concept itself). Nice to see _e.g._ here:
https://edamontology.github.io/edam-browser/#topic_3678
https://edamontology.github.io/edam-browser/#format_1975
even if there are more layers https://edamontology.github.io/edam-browser/#topic_0605

If you would like to dive into fixing this @HagerDakroury you'll be the hero! 🦸‍♀️
I'd personally suggest 3 options (but let's hear what @bryan-brancotte & @hmenager say)
- Simple: Just summing up the numbers, and showing something like _**upto X-times with its sub-concepts**_ (not caring that some might be counted multiple times)
- Or a correct solution (but I'm wondering if will perform well in case of 1000s of entries found): Retrieving the full lists for each sub-concept, creating a union set (_i.e._ no duplicates), and counting them.
- Maybe a compromise: Like above but cut off if e.g. more than 1000 and display just >1000. Or a smaller number if needed.

**Comment by HagerDakroury** [comment]
> 
> In any case, I found out where the numbers come from! 🙌🏽
> They are just the actual number for the given concept, multiplied by the number of its sub-concepts +1 (the concept itself). 
> 

Impressive! 😃 I spent way too much time trying to crack the maths here. Can't believe it's that simple!

> If you would like to dive into fixing this @HagerDakroury you'll be the hero! 🦸‍♀️

Absolutely! I'd love to dig deep here. It'd be an interesting challenge.

> However, if the 2nd or 3rd solutions are too complex to implement, then we perhaps don't want anyone spending too much time with it, as it really should be the _bio.tools_ supplying those numbers from its API (hopefully one day, finally😊). What do you all think?

It'd really depend on how the tree is modeled and what's the data pulled from bio.tools right now (I haven't examined this part's code in detail yet)

I guess the 2 points to be figured out:

1) Are the children's counts even accessed right now? ( or the number of children is the only info gathered?)
2) Does it make more sense to make several recursive requests to bio.tools whenever a node is selected?
 or pull all the data from the beginning and just do the aggregation locally for each selected node.

Yes, it may take time to fix it, but sure would be thrilling to try!

> I'd personally suggest 3 options (but let's hear what @bryan-brancotte & @hmenager say)

I'll try exploring the code more in the meantime while waiting for their input.

**Comment by matuskalas** [comment]
Super cool, I like your attitude @HagerDakroury 🙌🏽🙌🏽🙌🏽
Awesome that you want to take it up as a challenge!! 😉
> I'll try exploring the code more in the meantime while waiting for their input.

Perfect! Million thanks & all best 🤞🏽

**Comment by bryan-brancotte** [comment]
Hi @HagerDakroury @matuskalas 

Thanks to both of you for diving in this issue, and thanks even more for finding example that helped me to understand what was going on. I added comments with 5275c73. The issue was that `get_api_url(queue[j].text)` was called with the value of an unset attribut in https://github.com/edamontology/edam-browser/blame/5275c73472659b67ebccfe05277a63044328adaa/js/bio.tools.api.js#L60 as it was not set, `get_api_url` was using the uri of the current (selected) element in https://github.com/edamontology/edam-browser/blame/5275c73472659b67ebccfe05277a63044328adaa/js/bio.tools.api.js#L135 which produced in the end that descendant usage was `term usage` * ` number of its sub-concepts +1`

[Patch](https://github.com/edamontology/edam-browser/commit/52944fd20e1d5974bc5a8f789c33ec16e226ceb9#diff-a79ee75d4c5cd598a91021ada789f969201530ecf2d3c5232dd560eed56a8b3eR60) is pushed so actual id of descendant are used, and as you can see it now works
![Capture d’écran de 2021-04-08 15-30-00](https://user-images.githubusercontent.com/11556772/114035389-58062b80-987f-11eb-8152-7926f41b36b8.png)

great team works !

**Comment by HagerDakroury** [comment]
Awesome! 👏  It's such a relief that it only took one line to fix that 😃 

Also, thanks for adding the comments! That made navigating the code so much easier.

**Comment by matuskalas** [comment]
Nice, thanks so much @bryan-brancotte for fixing this!! 🙌🏽

Is it then the 1st option from
> - Simple: Just summing up the numbers, and showing something like _**upto X-times with its sub-concepts**_ (not caring that some might be counted multiple times)
> - Or a correct solution (but I'm wondering whether it will perform well in case of 1000s of entries found): Retrieving the full lists for each sub-concept, creating a union set (_i.e._ no duplicates), and counting them.
> - Maybe a compromise: Like above but cut off if e.g. more than 1000 and display just >1000. Or a smaller number if needed.

?

Then it would be nice to slightly update the tooltip to include the "upto", wouldn't it?

And @HagerDakroury @bryan-brancotte @hmenager, do you think that a more sophisticated solution would be too slow or too complicated? I'd personally say that if it should be potentially slow or potentially a complex code hard to maintain, then let's be happy with the 1st "upto" solution 😊

**Comment by bryan-brancotte** [comment]
_re-opening as some question are unanswered_

**Comment by bryan-brancotte** [comment]
Thanks @matuskalas, and also for remininding me that for now it still count duplicated child :/ 

I would definitely go for option 2: show to correct sum. Note that we do not count the descendant when looking at node of depth 0 or 1, i.e: EDAM and the four root node.

I pushed fa66e07 where duplicated node are not counted twice

**Comment by HagerDakroury** [comment]
Removing duplicates is doable, I think fa66e07 took care of it and now it's the correct sum.

> And @HagerDakroury @bryan-brancotte @hmenager, do you think that a more sophisticated solution would be too slow or too complicated? I'd personally say that if it should be potentially slow or potentially a complex code hard to maintain, then let's be happy with the 1st "upto" solution 😊

I'd say it's slow but not unnecessarily slow 😄

Maybe the way to make this faster is by somehow saving the values returned for future reuse or doing the requests beforehand. But the performance is currently not that bad IMO. So, restructuring the calls is not a priority? Since they work perfectly fine now.

**Comment by matuskalas** [comment]
Very fast @bryan-brancotte, respect! 🥇 

It definitely looks like running fast & with some "caching" ☺

Do I understand it right that now it takes into account when the same sub-concept appears multiple times in the sub-"tree"?

There's another source of duplication I meant, between entries, and that one is perhaps the slow|complex one to solve?
Looking at the 2 sub-concepts of https://edamontology.github.io/edam-browser/#topic_0082, some tools in _bio.tools_ are annotated with both:
![Screen Shot 04-08-21 at 05 03 PM](https://user-images.githubusercontent.com/992660/114050279-6c045a00-988c-11eb-9ab6-23e63f385d3a.PNG)
![Screen Shot 04-08-21 at 04 51 PM](https://user-images.githubusercontent.com/992660/114049830-12039480-988c-11eb-8373-11c8d600e979.PNG) ![Screen Shot 04-08-21 at 04 51 PM 001](https://user-images.githubusercontent.com/992660/114049844-13cd5800-988c-11eb-912b-359a08b24b3a.PNG)

**Comment by bryan-brancotte** [comment]
Prior to https://github.com/edamontology/edam-browser/issues/45#issuecomment-815898424

Indeed fa66e07 toke care of it, but might not be optimal on the way to sum the dictionary. I also think performance are not that bad, but there are still way of improvement :
 * make bio.tools do the works :wink: @matuskalas 
 * improve the dictionary sum
 * do not call the api twice when a node is duplicated as fa66e07 only prevent that it is counted twice, 

After https://github.com/edamontology/edam-browser/issues/45#issuecomment-815898424

Sadly HOPMA will be counted twice, we only get the "count" attribut for a given concept. We might be able to actually get the tools, and then count them, but it would be better (cpu, network) to do it on bio.tools side.

**Comment by bryan-brancotte** [comment]
Just to write it somewhere, we store the information of whether a node is duplicated or not in `node.duplicate`, it is set [here](https://github.com/edamontology/edam-browser/blob/main/js/tree-reusable-d3.js#L401-L406)

**Comment by HagerDakroury** [comment]
> There's another source of duplication I meant, between entries, and that one is perhaps the slow|complex one to solve?
> Looking at the 2 sub-concepts of https://edamontology.github.io/edam-browser/#topic_0082, some tools in _bio.tools_ are annotated with both:
> 

Ohh, well-spotted! and definitely makes things harder :(

> * do not call the api twice when a node is duplicated as [fa66e07](https://github.com/edamontology/edam-browser/commit/fa66e07aadf864a584275be545fcd34498dc321d) only prevent that it is counted twice,
> 

Is this a common occurrence? (a node being duplicated). If not and since @matuskalas meant another duplication, it may not be a priority to optimize that.

> 
> Sadly HOPMA will be counted twice, we only get the "count" attribut for a given concept. We might be able to actually get the tools, and then count them, but it would be better (cpu, network) to do it on bio.tools side.

Definitely, it may be easy code-wise but a huge bump in the time complexity.

But if the accuracy is currently vital for your users, then an upto indication may be necessary after all.

**Comment by bryan-brancotte** [comment]
Hi, I just re-read the labels, we never say that there is 58 tools associated with [Splicing analysis](https://edamontology.github.io/edam-browser/#operation_2499) for example, we say that it is used 58 times, and  `192 times with its 2 descendants`. The count are thus valid, they count the usage and we indicate it.

We could clarify and indicate that there is `192 annotation with the concept and its 2 descendants`

**Comment by bryan-brancotte** [comment]
@matuskalas as you were the one spotting that the count is ~~incorrect~~ ambiguous, do we change the text, or keep it as is and close the issue ?
